compiler/codegen/CodeGenPrep.cpp

```diff
@@ -177,6 +179,125 @@ OMR::CodeGenerator::lowerTreesPreChildrenVisit(TR::Node * parent, TR::TreeTop *
+   // of the value being constructed. The second child is
+   // the one that actually initializes the first field.
+   for (int i = 1; i < node->getNumChildren(); ++i)
+      {
+      auto* ttNode = TR::Node::create(TR::treetop, 1);
+      ttNode->setFirst(node->getChild(i));
+      prevTreeTop = TR::TreeTop::create(comp, ttNode, NULL, prevTreeTop);
+      prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+      }
+
+   // Create new TreeTop to anchor the new. The TreeTop for
+   // the newValue cannot be reused because it is still used
+   // just after lowering the node.
+   auto* newNode = TR::Node::createWithSymRef(TR::New, 1, node->getSymbolReference());
+   newNode->setFirst(node->getChild(0));
+   auto* ttNode = TR::Node::create(TR::treetop, 1);
+   ttNode->setAndIncChild(0, newNode);
+   prevTreeTop = TR::TreeTop::create(comp, ttNode, NULL, prevTreeTop);
+   prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+
+   // Get the class and layout of the value type
+   auto* valueClass = static_cast<TR_OpaqueClassBlock *>(newNode->getSymbol()->getStaticSymbol()->getStaticAddress());
+   const TR::TypeLayout* typeLayout = comp->typeLayout(valueClass);
+
+   // Iterate over every child that sets a field and generated
+   // a store for the value.
+   for (int i = 1; i < node->getNumChildren(); ++i)
+      {
+      TR::IL il;
+      auto storeOpCode = il.opCodeForIndirectStore(node->getChild(i)->getDataType());
+      const TR::TypeLayoutEntry fieldEntry = typeLayout->entry(i - 1);
+      auto* symref = comp->getSymRefTab()->findOrFabricateShadowSymbol(valueClass,
+                                                                       fieldEntry._datatype,
+                                                                       fieldEntry._offset,
+                                                                       fieldEntry._isVolatile,
+                                                                       fieldEntry._isPrivate,
+                                                                       fieldEntry._isFinal,
+                                                                       fieldEntry._fieldname,
+                                                                       fieldEntry._typeSignature
+                                                                       );
+      auto* storeNode = TR::Node::createWithSymRef(storeOpCode, 2, symref);
+      storeNode->setAndIncChild(0, newNode);
+      storeNode->setAndIncChild(1, node->getChild(i));
+      prevTreeTop = TR::TreeTop::create(comp, storeNode, NULL, prevTreeTop);
+      prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+      }
+
+   // add memory fence
+   prevTreeTop = TR::TreeTop::create(comp, TR::Node::create(TR::fence, 0), NULL, prevTreeTop);
+   prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+
+   // finish by linking to the next TreeTop
+   prevTreeTop->setNextTreeTop(tt->getNextRealTreeTop());
```

Comment:
Is the use of `getNextRealTreeTop` here intentional? It seems to me like this should be using `getNextTreeTop` instead:

```suggestion
   prevTreeTop->setNextTreeTop(tt->getNextTreeTop());
```

Reply:
Yes, it should be. Fixed.